labgrid/driver/qemudriver.py

```diff
@@ -0,0 +1,164 @@
+"""The qemudriver implements a driver to use a qemu target"""
+import logging
+import select
+import shlex
+import shutil
+import socket
+import subprocess
+import tempfile
+
+import attr
+from pexpect import TIMEOUT
+
+from ..factory import target_factory
+from ..protocol import PowerProtocol, ConsoleProtocol
+from ..step import step
+from .common import Driver
+from .consoleexpectmixin import ConsoleExpectMixin
+from ..util.qmp import QMPMonitor
+
+
+@target_factory.reg_driver
+@attr.s
+class QemuDriver(ConsoleExpectMixin, Driver, PowerProtocol, ConsoleProtocol):
+    """
+    The QemuDriver implements an interface to start targets as qemu instances.
+
+    Args:
+        qemu_bin (str): Path to the qemu binary
+        machine (str): qemu machine type
+        cpu (str): qemu cpu type
+        memory (str): qemu memory size (ends with M or G)
+        boot_args (str): kernel boot argument
+        extra_args (str): extra qemu arguments, are passed directly to the qemu binary
+        kernel (str): path to the kernel image
+        rootfs (str): path to the rootfs for the virtio-9p filesystem
+        dtb (str): optional, path to the compiled device tree
+    """
+    qemu_bin = attr.ib(validator=attr.validators.instance_of(str))
+    machine = attr.ib(validator=attr.validators.instance_of(str))
+    cpu = attr.ib(validator=attr.validators.instance_of(str))
+    memory = attr.ib(validator=attr.validators.instance_of(str))
+    boot_args = attr.ib(validator=attr.validators.instance_of(str))
+    extra_args = attr.ib(validator=attr.validators.instance_of(str))
+    kernel = attr.ib(
+        validator=attr.validators.optional(attr.validators.instance_of(str)))
+    rootfs = attr.ib(
+        validator=attr.validators.optional(attr.validators.instance_of(str)))
+    dtb = attr.ib(
+        default=None,
+        validator=attr.validators.optional(attr.validators.instance_of(str)))
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.logger = logging.getLogger("{}:".format(self))
+        self.status = 0  #pylint: disable=attribute-defined-outside-init
+        self._cmd = [self.qemu_bin]
```

Comment:
Move the _cmd construction to on_activate, otherwise it will be extended every time this driver is activated.